File: backend/pkg/gameplugin/sdk/game-plugin/src/transport_stdio.ts
import { Envelope } from './protocol';

export const STDIO_MAX_FRAME_SIZE = 16 * 1024 * 1024;
export const STDIO_FRAME_HEADER_SIZE = 4;
export const STDIO_BINARY_FRAME_FLAG = 0x80000000;
export const STDIO_FRAME_LENGTH_MASK = 0x7fffffff;

export interface StdioTransportOptions {
  stdin?: NodeJS.ReadableStream;
  stdout?: NodeJS.WritableStream;
  maxFrameSize?: number;
}

export class StdioTransport {
  private stdin: NodeJS.ReadableStream;
  private stdout: NodeJS.WritableStream;
  private maxFrameSize: number;
  private buffer: Buffer | null = null;
  private resolveReceive: ((data: Buffer) => void) | null = null;

  constructor(options: StdioTransportOptions = {}) {
    this.stdin = options.stdin ?? process.stdin;
    this.stdout = options.stdout ?? process.stdout;
    this.maxFrameSize = options.maxFrameSize ?? STDIO_MAX_FRAME_SIZE;
    this.setupReadHandler();
  }

  private setupReadHandler(): void {
    this.stdin.on('data', (chunk: Buffer) => {
      if (this.resolveReceive) {
        this.resolveReceive(chunk);
        this.resolveReceive = null;
      } else {
        this.buffer = this.buffer
          ? Buffer.concat([this.buffer, chunk])
          : chunk;
      }
    });
  }

  private async readFromBuffer(size: number): Promise<Buffer | null> {
    if (this.buffer && this.buffer.length >= size) {
      const result = this.buffer.subarray(0, size);
      this.buffer = this.buffer.subarray(size);
      if (this.buffer.length === 0) {
        this.buffer = null;
      }
      return result;
    }

    return new Promise<Buffer | null>((resolve) => {
      this.resolveReceive = (chunk: Buffer) => {
        const combined = this.buffer
          ? Buffer.concat([this.buffer, chunk])
          : chunk;
        if (combined.length >= size) {
          const result = combined.subarray(0, size);
          this.buffer = combined.subarray(size);
          if (this.buffer.length === 0) {
            this.buffer = null;
          }
          resolve(result);
        } else {
          this.buffer = combined;
          resolve(null);
        }
      };
    });
  }

  async send(message: Envelope): Promise<void> {
    const data = JSON.stringify(message);
    const payload = Buffer.from(data, 'utf-8');

    if (payload.length > this.maxFrameSize) {
      throw new Error(`frame size ${payload.length} exceeds limit ${this.maxFrameSize}`);
    }

    const header = Buffer.alloc(STDIO_FRAME_HEADER_SIZE);
    header.writeUInt32BE(payload.length, 0);

    return new Promise<void>((resolve, reject) => {
      this.stdout.write(Buffer.concat([header, payload]), (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }


  async sendBinaryFrame(message: Envelope, objectId: string, offset: number, data: Uint8Array): Promise<void> {
    if (!message.id) throw new Error('binary frame requires message id');
    if (!objectId || !objectId.startsWith('bin_')) throw new Error('binary frame requires valid object id');
    if (!Number.isSafeInteger(offset) || offset < 0) throw new Error('binary frame offset must be a non-negative safe integer');
    if (!(data instanceof Uint8Array) || data.byteLength === 0) throw new Error('binary frame payload must not be empty');

    const frameHeader = {
      protocol: message.protocol,
      id: message.id,
      runtimeId: message.runtimeId,
      pluginId: message.pluginId,
      serviceId: message.serviceId,
      generation: message.generation,
      objectId,
      offset,
    };
    const headerJson = Buffer.from(JSON.stringify(frameHeader), 'utf-8');
    const payload = Buffer.from(data);
    const bodyLength = 4 + headerJson.length + payload.length;
    if (bodyLength <= 4 || bodyLength > this.maxFrameSize || bodyLength > STDIO_FRAME_LENGTH_MASK) {
      throw new Error(`binary frame size ${bodyLength} exceeds limit ${this.maxFrameSize}`);
    }

    const outer = Buffer.alloc(STDIO_FRAME_HEADER_SIZE);
    outer.writeUInt32BE((STDIO_BINARY_FRAME_FLAG | bodyLength) >>> 0, 0);
    const meta = Buffer.alloc(4);
    meta.writeUInt32BE(headerJson.length, 0);
    const frame = Buffer.concat([outer, meta, headerJson, payload]);

    return new Promise<void>((resolve, reject) => {
      this.stdout.write(frame, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  async receive(): Promise<Envelope> {
    const headerBuf = await this.readExact(STDIO_FRAME_HEADER_SIZE);
    const frameLen = headerBuf.readUInt32BE(0);

    if (frameLen === 0) {
      throw new Error('invalid frame: zero length');
    }
    if (frameLen > this.maxFrameSize) {
      throw new Error(`frame size ${frameLen} exceeds limit ${this.maxFrameSize}`);
    }

    const data = await this.readExact(frameLen);
    const jsonStr = data.toString('utf-8');
    return JSON.parse(jsonStr) as Envelope;
  }

  private async readExact(size: number): Promise<Buffer> {
    while (true) {
      const result = await this.readFromBuffer(size);
      if (result) return result;
      await new Promise((resolve) => setImmediate(resolve));
    }
  }

  async close(): Promise<void> {
    return Promise.resolve();
  }
}

File: backend/pkg/gameplugin/sdk/go/transport_stdio.go
package sdk

import (
	"context"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sync"

	"github.com/u-ai/backend/pkg/gameplugin/protocol"
)

const (
	StdioMaxFrameSize    = 16 * 1024 * 1024
	StdioFrameHeaderSize = 4
	StdioBinaryFrameFlag = uint32(1 << 31)
	stdioFrameLengthMask = ^StdioBinaryFrameFlag
)

type StdioTransport struct {
	reader   io.Reader
	writer   io.Writer
	closer   io.Closer
	writerMu sync.Mutex
	maxSize  int64
}

type StdioTransportConfig struct {
	Reader       io.Reader
	Writer       io.Writer
	Closer       io.Closer
	MaxFrameSize int64
}

func NewStdioTransport(config StdioTransportConfig) *StdioTransport {
	maxSize := config.MaxFrameSize
	if maxSize <= 0 {
		maxSize = StdioMaxFrameSize
	}
	reader := config.Reader
	if reader == nil {
		reader = os.Stdin
	}
	writer := config.Writer
	if writer == nil {
		writer = os.Stdout
	}
	return &StdioTransport{
		reader:  reader,
		writer:  writer,
		closer:  config.Closer,
		maxSize: maxSize,
	}
}

func NewDefaultStdioTransport() *StdioTransport {
	return NewStdioTransport(StdioTransportConfig{})
}

func (t *StdioTransport) Send(ctx context.Context, msg protocol.Envelope) error {
	if err := ctx.Err(); err != nil {
		return err
	}

	data, err := json.Marshal(msg)
	if err != nil {
		return fmt.Errorf("marshal envelope failed: %w", err)
	}

	if int64(len(data)) > t.maxSize {
		return fmt.Errorf("frame size %d exceeds limit %d", len(data), t.maxSize)
	}

	select {
	case <-ctx.Done():
		return ctx.Err()
	default:
	}

	t.writerMu.Lock()
	defer t.writerMu.Unlock()

	if err := binary.Write(t.writer, binary.BigEndian, uint32(len(data))); err != nil {
		return fmt.Errorf("write frame header failed: %w", err)
	}

	if _, err := t.writer.Write(data); err != nil {
		return fmt.Errorf("write frame data failed: %w", err)
	}

	return nil
}

func (t *StdioTransport) SendBinaryFrame(ctx context.Context, message protocol.Envelope, objectID string, offset int64, data []byte) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	if len(data) == 0 {
		return fmt.Errorf("binary frame payload must not be empty")
	}
	header := protocol.BinaryFrameHeader{
		Protocol:   message.Protocol,
		ID:         message.ID,
		RuntimeID:  message.RuntimeID,
		PluginID:   message.PluginID,
		ServiceID:  message.ServiceID,
		Generation: message.Generation,
		ObjectID:   objectID,
		Offset:     offset,
	}
	if err := header.Validate(); err != nil {
		return err
	}
	headerJSON, err := json.Marshal(header)
	if err != nil {
		return fmt.Errorf("marshal binary frame header failed: %w", err)
	}
	bodyLen := 4 + len(headerJSON) + len(data)
	if int64(bodyLen) > t.maxSize || uint64(bodyLen) > uint64(stdioFrameLengthMask) {
		return fmt.Errorf("binary frame size %d exceeds limit %d", bodyLen, t.maxSize)
	}

	t.writerMu.Lock()
	defer t.writerMu.Unlock()
	frameHeader := make([]byte, 4)
	binary.BigEndian.PutUint32(frameHeader, StdioBinaryFrameFlag|uint32(bodyLen))
	metaHeader := make([]byte, 4)
	binary.BigEndian.PutUint32(metaHeader, uint32(len(headerJSON)))
	for _, part := range [][]byte{frameHeader, metaHeader, headerJSON, data} {
		if _, err := t.writer.Write(part); err != nil {
			return fmt.Errorf("write binary frame failed: %w", err)
		}
	}
	if flusher, ok := t.writer.(interface{ Flush() error }); ok {
		if err := flusher.Flush(); err != nil {
			return fmt.Errorf("flush writer failed: %w", err)
		}
	}
	return nil
}

func (t *StdioTransport) Receive(ctx context.Context) (protocol.Envelope, error) {
	if err := ctx.Err(); err != nil {
		return protocol.Envelope{}, err
	}

	headerBuf := make([]byte, StdioFrameHeaderSize)

	if err := t.readFull(headerBuf); err != nil {
		if err == io.EOF || err == io.ErrUnexpectedEOF {
			return protocol.Envelope{}, io.EOF
		}
		return protocol.Envelope{}, fmt.Errorf("read frame header failed: %w", err)
	}

	frameLen := binary.BigEndian.Uint32(headerBuf)
	if frameLen == 0 {
		return protocol.Envelope{}, fmt.Errorf("invalid frame: zero length")
	}
	if int64(frameLen) > t.maxSize {
		return protocol.Envelope{}, fmt.Errorf("frame size %d exceeds limit %d", frameLen, t.maxSize)
	}

	data := make([]byte, frameLen)
	if err := t.readFull(data); err != nil {
		if err == io.EOF || err == io.ErrUnexpectedEOF {
			return protocol.Envelope{}, io.EOF
		}
		return protocol.Envelope{}, fmt.Errorf("read frame data failed: %w", err)
	}

	var envelope protocol.Envelope
	if err := json.Unmarshal(data, &envelope); err != nil {
		return protocol.Envelope{}, fmt.Errorf("unmarshal envelope failed: %w", err)
	}

	return envelope, nil
}

func (t *StdioTransport) readFull(buf []byte) error {
	_, err := io.ReadFull(t.reader, buf)
	return err
}

func (t *StdioTransport) Close() error {
	t.writerMu.Lock()
	defer t.writerMu.Unlock()

	if t.closer == nil {
		return nil
	}
	return t.closer.Close()
}
